### PR TITLE
allow local NPC functions to be public or private

### LINKED ISSUE #2137: local functions can be called as events
since local function names and labels are both stored in the same strdb it is possible to call a local function as if it were an event when it has a name that begins with `On`
```c
// define a function in MyNPC
function OnMyFunction {
	debugmes("hello!");
	return;
}

// then it can be called from anywhere:
donpcevent("MyNPC::OnMyFunction");
```

```ini
[Debug]: script debug : 0 110000354 : hello!
[Warning]: script:run_func: return without callfunc or callsub!
```
This affects not just custom events but rather every single event calls (oninit, onlogin, mob death, timers, clock, woe, …) since they all use the same functions

<br>

---
We should either make it illegal or officially support it, which would be a great idea as we could get rid of most labels and use functions everywhere. As syntax for the event string we could use
```c++
MyNPC::MyFunction()
```
so if parentheses are found it is a function call, else it is a normal label call

### PR DESCRIPTION
**Pull request by Helianthella**
This is my proposal to fix #2137:

* local functions can no longer be called as events, even if their label begins with `On`
* local functions may be explicitly exported by defining them as `public function ...`.
this allows other scripts to call them as if they were part of their own script
* exported local functions can be called from any script by doing `"npc name"::MyFunction();`

<br>

```c++
-	script	npc1	FAKE_NPC,{

	private function NonExported {
		return "[1] non-exported";
	}

	public function Exported {
		return "[1] exported";
	}

	// local call:
	debugmes(NonExported()); // => [1] non-exported
	debugmes(Exported()); // => [1] exported

	// call in another NPC:
	debugmes("npc2"::NonExported()); // ERROR: not exported
	debugmes("npc2"::Exported()); // => [2] exported
}


-	script	npc2	FAKE_NPC,{

	function NonExported {
		return "[2] non-exported";
	}

	public function Exported {
		return "[2] exported";
	}

	// local call:
	debugmes(NonExported()); // => [2] non-exported
	debugmes(Exported()); // => [2] exported

	// call in another NPC:
	debugmes("npc1"::NonExported()); // ERROR: not exported
	debugmes("npc1"::Exported()); // => [1] exported
}
```

```c++
// local call syntax (unchanged):
<function>({<arg>...})

// public call syntax:
"<npc name>"::<function>({<arg>...})

// also:
callfunctionofnpc("<npc name>", "<function>"{, <arg>...})
```

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/2142)

File: src/map/script.c [hunk 1]
```diff
@@ -1903,8 +1943,15 @@ static const char *parse_syntax(const char *p)
 		} else if( p2 - p == 8 && strncmp(p, "function", 8) == 0 ) {
 			// internal script function
 			const char *func_name;
+			bool is_extern = false;
 
 			func_name = script->skip_space(p2);
+
+			if (*func_name == ':') {
```
Comment (mekolat) [inline src/map/script.c]: to mark functions as exported I chose `:` because it is short and easy to remember, but we could use a keyword if this is preferred, such as:
* `extern function MyFunction { … }`
* `export function MyFunction { … }`
* `public function MyFunction { … }`

File: src/map/npc.h [hunk 2]
```diff
@@ -55,6 +55,7 @@ struct npc_timerevent_list {
 struct npc_label_list {
 	char name[NAME_LENGTH];
 	int pos;
+	uint8 flags; // is_extern, is_userfunc
```
Comment (4144) [inline src/map/npc.h]: probably better here use enum?

File: src/map/npc.c [hunk 3]
```diff
@@ -387,7 +387,7 @@ static int npc_event_export(struct npc_data *nd, int i)
 	Assert_ret(i >= 0 && i < nd->u.scr.label_list_num);
 	lname = nd->u.scr.label_list[i].name;
 	pos = nd->u.scr.label_list[i].pos;
-	if ((lname[0] == 'O' || lname[0] == 'o') && (lname[1] == 'N' || lname[1] == 'n')) {
+	if ((nd->u.scr.label_list[i].flags & 0x1) && !(nd->u.scr.label_list[i].flags & 0x2)) {
```
Comment (4144) [inline src/map/npc.c]: 0x1 and 0x2 probably better replace to enum values?

**Comment by 4144** [comment]
and i think need add some tests to call local functions in CI

File: src/map/npc.h (hunk 2 above)
Comment (mekolat) [reply src/map/npc.h]: right, sorry

**Comment by mekolat** [comment]
@4144 yes, sure, I will do this after I write the docs

File: doc/script_commands.txt [hunk 4]
```diff
@@ -1963,18 +1967,31 @@ prontera,150,150,0	script	TestNPC	123,{
...
-}
 
 
 ---------------------------------------
 
+*<function name>({<arg>...})
+*"<npc name>"::<function name>({<arg>...})
+
+In its first form, calls a previously defined local function. In its second
+form, calls a previously defined exported local function of another NPC.
```
Comment (mekolat) [inline doc/script_commands.txt]: `calls a previously defined exported local function of another NPC` … that's a mouthful,
but I don't really know how to explain it in simpler terms

File: doc/script_commands.txt [hunk 5]
```diff
@@ -1904,57 +1904,61 @@ will result in error and termination of the script.
 ---------------------------------------
 
 *function <function name>;
-*<function name>{(<argument>, ...<argument>)};
```
Comment (mekolat) [inline doc/script_commands.txt]: something here bothers me: we're using curly braces to denote optional arguments, but in this case literal curly braces are required; how should this be written? escaped curly braces (`\{ <code> \}`) or a custom sequence like we do for `%TAB%`?

File: doc/script_commands.txt [hunk 6]
```diff
@@ -1904,57 +1904,61 @@ will result in error and termination of the script.
...
 *function <function name> {
```
Comment (mekolat) [inline doc/script_commands.txt]: something here bothers me: we're using curly braces to denote optional arguments, but in this case literal curly braces are required; how should this be written? escaped curly braces (`\{ <code> \}`) or a custom sequence like we do for `%TAB%`?

Comment (MishimaHaruna) [reply doc/script_commands.txt]: I think escaping them might cause more confusion, so I'd advise against that (or someone will really try to write it as `\{` in scripts).

I'd personally leave it as is (I don't believe it's ambiguous because it's not a function argument). Another option would be `%LCURLY%` / `%RCURLY%` but considering that I already find the `%TAB%` annoyingly hard to read...

**Comment by mekolat** [comment]
@4144 I have added unit tests

**Comment by mekolat** [comment]
moving this to the next milestone

**Comment by Helianthella** [comment]
- Updated to use a `public` / `private` prefix instead of a colon
- Renamed the `foreigncall()` buildin that is used internally to `callfunctionofnpc()`
- Updated the documentation and also documented `callfunctionofnpc()`

File: src/map/script.c [hunk 7]
```diff
@@ -848,57 +848,92 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
...
-			if( *arg != '*' ) ++ arg;
-		}
+			} else {;
+				script->addl(script->buildin_callfunc_ref);
+				script->addc(C_ARG);
+				script->addc(C_STR);
+				while( *name ) script->addb(*name ++);
+				script->addb(0);
+				arg = script->buildin[script->str_data[script->buildin_callfunc_ref].val];
+				if( *arg != '*' ) ++ arg;
```
Comment (Kenpachi2k13) [inline src/map/script.c:933]: Misplaced spaces and `++arg;` should be placed in new line.

File: src/map/script.c [hunk 8]
```diff
@@ -1240,6 +1275,13 @@ static const char *parse_simpleexpr(const char *p)
 	} else if (is_number(p)) {
 		return script->parse_simpleexpr_number(p);
 	} else if(*p == '"') {
+		const char *p2 = p + 1;
+		while (*p2 != '"') ++p2;
```
Comment (Kenpachi2k13) [inline src/map/script.c:1279]: `++p2;` should be placed in new line.

File: src/map/script.c [hunk 9]
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
...
+
+	if (*p2 == ';') {
+		// function <name> ;
+		// function declaration - just register the name
+		int l = script->add_word(func_name);
+
+		if (script->str_data[l].type == C_NOP) {
+			// register only, if the name was not used by something else
+			script->str_data[l].type = C_USERFUNC;
+		} else if(script->str_data[l].type == C_USERFUNC) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:1641]: Missing space after `if`.

File: src/map/script.c [hunk 10]
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
...
+			;  // already registered
+		} else {
+			disp_error_message("parse_syntax:function: function name is already in use", func_name);
+		}
+
+		// Close condition of if, for, while
+		p = script->parse_syntax_close(p2 + 1);
+		return p;
+	} else if(*p2 == '{') {
```
Comment (Kenpachi2k13) [inline src/map/script.c:1650]: Missing space after `if`.

File: src/map/script.c [hunk 11]
```diff
@@ -1577,6 +1619,74 @@ static const char *parse_line(const char *p)
...
+
+		// Set the position of the function (label)
+		int l = script->add_word(func_name);
+
+		if (script->str_data[l].type == C_NOP || script->str_data[l].type == C_USERFUNC) {
+			// register only, if the name was not used by something else
+			script->str_data[l].type = C_USERFUNC;
+			script->set_label(l, VECTOR_LENGTH(script->buf), p);
+
+			if (script->parse_options & SCRIPT_USE_LABEL_DB) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:1674]: Please use explicit comparison for non-boolean expressions.

File: src/map/script.c [hunk 12]
```diff
@@ -2674,11 +2746,14 @@ static struct script_code *parse_script(const char *src, const char *file, int l
 			disp_error_message("unexpected end of script",p);
 		// Special handling only label
 		tmpp=script->skip_space(script->skip_word(p));
-		if(*tmpp==':' && !(strncmp(p,"default:",8) == 0 && p + 7 == tmpp)) {
+		if(*tmpp==':' && !(strncmp(p,"default:",8) == 0 && p + 7 == tmpp) &&
```
Comment (Kenpachi2k13) [inline src/map/script.c:2749]: Missing spaces and please break the line before the logical operator, not after.

File: src/map/script.c [hunk 13]
```diff
@@ -2674,11 +2746,14 @@ static struct script_code *parse_script(const char *src, const char *file, int l
...
+			!(strncmp(p, "function", 8) == 0 && script->skip_space(p + 8) == tmpp)) {
 			i=script->add_word(p);
 			script->set_label(i, VECTOR_LENGTH(script->buf), p);
-			if( script->parse_options&SCRIPT_USE_LABEL_DB )
-				script->label_add(i, VECTOR_LENGTH(script->buf));
+			if (script->parse_options & SCRIPT_USE_LABEL_DB) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:2753]: Please use explicit comparison for non-boolean expressions.

File: src/map/script.c [hunk 14]
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
...
+		ShowError("script:callfunctionofnpc: NPC not found.\n");
+		st->state = END;
+		return false;
+	}
+
+	for (i = 0; i < nd->u.scr.label_list_num; ++i) {
+		if (strcmp(nd->u.scr.label_list[i].name, function_name) == 0) {
+			if (nd->u.scr.label_list[i].flags & LABEL_IS_EXTERN && nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
+				pos = nd->u.scr.label_list[i].pos;
+			} else if (nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:6500]: Please use explicit comparison for non-boolean expressions.

File: src/map/script.c [hunk 15]
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
...
+
+	if (nd == NULL) {
+		ShowError("script:callfunctionofnpc: NPC not found.\n");
+		st->state = END;
+		return false;
+	}
+
+	for (i = 0; i < nd->u.scr.label_list_num; ++i) {
+		if (strcmp(nd->u.scr.label_list[i].name, function_name) == 0) {
+			if (nd->u.scr.label_list[i].flags & LABEL_IS_EXTERN && nd->u.scr.label_list[i].flags & LABEL_IS_USERFUNC) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:6498]: Please use explicit comparison for non-boolean expressions.

File: src/map/script.c [hunk 16]
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
...
+	script->push_retinfo(st->stack, ri, ref);
+
+	st->pos = pos;
+	st->script = nd->u.scr.script;
+	st->stack->defsp = st->stack->sp;
+	st->state = GOTO;
+	st->stack->scope.vars = i64db_alloc(DB_OPT_RELEASE_DATA);
+	st->stack->scope.arrays = idb_alloc(DB_OPT_BASE);
+
+	if (!st->script->local.vars) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:6549]: Please use explicit comparison for non-boolean expressions.

File: src/map/script.c [hunk 17]
```diff
@@ -6397,6 +6472,87 @@ static BUILDIN(callfunc)
...
 	return true;
 }
+
+static BUILDIN(callfunctionofnpc) {
+	struct npc_data *nd = NULL;
+	const char *function_name = script_getstr(st, 3);
+	struct script_retinfo *ri = NULL;
+	struct reg_db *ref = NULL;
+	int pos = -1;
+	int i = 0;
```
Comment (Kenpachi2k13) [inline src/map/script.c:6482]: Please fix scope of variables.

File: src/map/script.h [hunk 18]
```diff
@@ -952,6 +954,7 @@ struct script_interface {
 	const char * (*parse_syntax_close) (const char *p);
 	const char * (*parse_syntax_close_sub) (const char *p, int *flag);
 	const char * (*parse_syntax) (const char *p);
+	const char * (*parse_syntax_function) (const char *p, bool is_public);
```
Comment (Kenpachi2k13) [inline src/map/script.h:957]: Asterisk should be adjacent to the field name.

File: src/map/script.c [hunk 19]
```diff
@@ -848,57 +848,92 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
...
+				// buildin function
+				script->syntax.last_func = script->str_data[func].val;
+				script->addl(func);
+				script->addc(C_ARG);
+			}
+
+			arg = script->buildin[script->str_data[func].val];
+			if (script->str_data[func].deprecated)
+				DeprecationWarning(p);
+			if( !arg ) arg = &null_arg; // Use a dummy, null string
```
[the remark this answers is not in the document]
Comment (Helianthella) [reply src/map/script.c:908]: Most of the wrong formatting is actually just re-indented, unmodified legacy code but I don't mind refactoring it

File: src/map/script.c [hunk 20]
```diff
@@ -848,121 +848,198 @@ static const char *parse_callfunc(const char *p, int require_paren, int is_custo
...
 	}
-	if( arg && *arg && *arg != '?' && *arg != '*' )
-		disp_error_message2("parse_callfunc: not enough arguments, expected ','", p, script->config.warn_func_mismatch_paramnum);
-	if( script->syntax.curly[script->syntax.curly_count].type != TYPE_ARGLIST )
-		disp_error_message("parse_callfunc: DEBUG last curly is not an argument list",p);
-	if( script->syntax.curly[script->syntax.curly_count].flag == ARGLIST_PAREN ) {
-		if( *p != ')' )
-			disp_error_message("parse_callfunc: expected ')' to close argument list",p);
+
+	if (arg && *arg && *arg != '?' && *arg != '*') {
```
Comment (Kenpachi2k13) [inline src/map/script.c:1013]: `if (arg != NULL && *arg != 0 && ...`

File: src/map/script.c [hunk 21]
```diff
@@ -1577,6 +1667,74 @@ static const char *parse_line(const char *p)
...
+		int l = script->add_word(func_name);
+
+		if (script->str_data[l].type == C_NOP) {
+			// register only, if the name was not used by something else
+			script->str_data[l].type = C_USERFUNC;
+		} else if (script->str_data[l].type == C_USERFUNC) {
+			;  // already registered
+		} else {
+			disp_error_message("script:parse_syntax_function: function name is already in use", func_name);
+		}
```
Comment (Kenpachi2k13) [inline src/map/script.c:1693]: To avoid that empty block you could change `== C_USERFUNC` to `!= C_USERFUNC` and show the error message in this block and remove the `else` block completely.
(Having an additional block just to insert a comment brings no benefit in my opinion.)

File: src/map/script.c [hunk 22]
```diff
@@ -2668,25 +2788,32 @@ static struct script_code *parse_script(const char *src, const char *file, int l
 		}
 	}
 
-	while( script->syntax.curly_count != 0 || *p != end )
-	{
-		if( *p == '\0' )
-			disp_error_message("unexpected end of script",p);
+	while (script->syntax.curly_count != 0 || *p != end) {
+		if(*p == '\0') {
```
Comment (Kenpachi2k13) [inline src/map/script.c:2792]: Missing space.

File: src/map/script.c [hunk 23]
```diff
@@ -2668,25 +2788,32 @@ static struct script_code *parse_script(const char *src, const char *file, int l
...
+			disp_error_message("script:parse_script: unexpected end of script", p);
+		}
+
 		// Special handling only label
-		tmpp=script->skip_space(script->skip_word(p));
-		if(*tmpp==':' && !(strncmp(p,"default:",8) == 0 && p + 7 == tmpp)) {
-			i=script->add_word(p);
+		tmpp = script->skip_space(script->skip_word(p));
+
+		if (*tmpp==':' && !(strncmp(p, "default:", 8) == 0 && p + 7 == tmpp)
```
Comment (Kenpachi2k13) [inline src/map/script.c:2799]: Missing spaces around `==`.

File: src/map/script.c [hunk 24]
```diff
@@ -6397,6 +6524,89 @@ static BUILDIN(callfunc)
 
 	return true;
 }
+
+static BUILDIN(callfunctionofnpc) {
+	const char *function_name = script_getstr(st, 3);
```
Comment (Kenpachi2k13) [inline src/map/script.c:6529]: Fix scope. (Move down to `int pos = -1;`)

**Comment by Helianthella** [comment]
I'll do some commit squashing once the review process is over

**Comment by Helianthella** [comment]
I added two config flags for more flexibility:
- `functions_private_by_default` defines whether functions are private or public when not explicitly set as public or private.
- `functions_as_events` allows functions to be called as events if made public. This allows for backward compatibility if someone was using this bug as a feature.

```c
-	script	@mycommand	FAKE_NPC,{
	public function OnInit {
		bindatcmd("mycommand", "@mycommand::OnCall");
		return;
	}

	private function doSomething {
		dispbottom(getarg(0));
		return;
	}

	public function OnCall {
		doSomething(.@atcmd_parameters$[0]);
		return;
	}
}
```

File: src/map/script.c [hunk 25]
```diff
@@ -6397,6 +6530,89 @@ static BUILDIN(callfunc)
 
 	return true;
 }
+
+static BUILDIN(callfunctionofnpc) {
```
Comment (Kenpachi2k13) [inline src/map/script.c:6534]: Please add a documentation. (Switching between code and script_commands.txt is a pain in the arse. :smile:)

File: src/map/script.c [hunk 26]
```diff
@@ -1577,6 +1669,72 @@ static const char *parse_line(const char *p)
 	return p;
 }
 
+static const char *parse_syntax_function (const char *p, bool is_public)
```
Comment (Kenpachi2k13) [inline src/map/script.c:1672]: Please add a documentation.

File: src/map/map.h [hunk 27]
```diff
@@ -329,6 +329,11 @@ enum bl_type {
 
 enum npc_subtype { WARP, SHOP, SCRIPT, CASHSHOP, TOMB };
 
+enum script_label_flags {
```
Comment (Kenpachi2k13) [inline src/map/map.h:332]: Please add a documentation. (At least a small note about its purpose.)

**Comment by Kenpachi2k13** [comment]
Furthermore your commits need some squashing. :relaxed:

**Comment by Helianthella** [comment]
@Kenpachi2k13 as noted earlier, I do intend on squashing once the review process is done.
It makes it easier for maintainers to review new changes when the branch is not force-pushed
